### PR TITLE
ns for fx: decouple subgraph names from node names

### PR DESCRIPTION
**Pull request by vkuzo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52799 ns for fx: remove ".stats" suffix
* #52798 ns for fx: add node name and type to results
* #52789 ns for fx: make return type of ns APIs future proof
* #52779 ns for fx: unify return types of weight and activation APIs
* **#52771 ns for fx: decouple subgraph names from node names**
* #52681 ns for fx: update graph matching to handle dicts and tuples in node args
* #52402 ns for fx: update graph matching to not match nodes with equal types
* #52395 ns for fx: support linear_relu for weight matching
* #52368 ns for fx: allow graph matching of parents of cat

Summary:

Before this PR, subgraph names were derived from node names
in model B.  For example, if we had

```
A: linear0 -> relu0 -> ...
B: linear_relu0 -> ...
```

Then the subgraph name would be `linear_relu0`, and the outputs before this
PR would look like

```
{
  'linear_relu0': {
    'model_a': ...,
    'model_b': ...,
  },
}
```

This PR decouples subgraph naming from node names.
The outputs after this PR look like:

```
{
  # guaranteed to match the right subgraphs across different models
  # without needing more than one model during the prepare passes
  'base_op_torch.nn.functional.linear_0': {
    'model_a': ...,
    'model_b': ...,
  },
}
```

There are future requirements for which using node_name as subgraph name does not work well:
a. the need to support N models, without having all of them in memory at the same time
b. the need to support fusions and match subgraphs with related but non-equal types

This PR changes the naming of subgraphs to be based on two things:
1. the name of the underlying set of related ops (i.e. `torch.nn.functional.linear`)
2. the order in which this subgraph was named (i.e. `foo_0`, `foo_1`, ...)

Basically, we can't use a node name because of (a), since there must be
a reference model which node name other models must use, but that
reference model is not guaranteed to be available.  Note: we could add
some state and require the reference model to go through the APIs first,
saving the reference node names, but I'm deliberately not doing that
to minimize the state used throughout.

To support (b), we need a way to determine a name of a subgraph which is
the same for all related subgraphs (i.e. linear-relu vs quantized_linear
vs quantized_linear_relu). In this PR, this is done by using the base
aten op's name.  We use a string name so it looks nice in the output
(I tried `str(underlying_type)`, and it is not easy for humans to read).

Note: after this PR, it's hard to parse the results to see which layer
is related to which node in the graph. This will be fixed in a future PR
where we will store the node name on the logger, and expose it in the
output.

Test Plan:

```
python test/test_quantization.py TestFXGraphMatcher
python test/test_quantization.py TestFXGraphMatcherModels
python test/test_quantization.py TestFXNumericSuiteCoreAPIs
python test/test_quantization.py TestFXNumericSuiteCoreAPIsModels
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26652641](https://our.internmc.facebook.com/intern/diff/D26652641)

### CODE REVIEW

File: torch/quantization/ns/graph_matcher.py [hunk 1]
```diff
@@ -34,38 +31,45 @@ def get_type_a_related_to_b() -> Set[Tuple[Callable, Callable]]:
...
+def get_type_a_related_to_b(
+    base_name_to_sets_of_related_ops: Dict[str, Set[Callable]],
+) -> Set[Tuple[Callable, Callable]]:
+    # TODO(future PR): allow customizations
+    # TODO(future PR): reuse existing quantization mappings
+    # TODO(future PR): add the rest of modules and ops here
     type_a_related_to_b: Set[Tuple[Callable, Callable]] = set()
 
-    for s in sets_of_related_ops:
+    for base_name, s in base_name_to_sets_of_related_ops.items():
```
Comment (hx89) [inline torch/quantization/ns/graph_matcher.py:72]: nit: use "_" since "base_name" is not used?

File: torch/quantization/ns/graph_matcher.py [hunk 2]
```diff
@@ -280,18 +284,58 @@ def _get_node_relationship_type(
             return NodeTypeRelationship.NOT_RELATED
     return NodeTypeRelationship.NOT_RELATED
 
-def _get_name_for_subgraph_pair(
+def _get_name_for_subgraph(
     start_node_a: Node,
     end_node_a: Node,
-    start_node_b: Node,
-    end_node_b: Node,
+    gm_a: GraphModule,
```
Comment (hx89) [inline torch/quantization/ns/graph_matcher.py:290]: nit: Since there's no model b now, may just use name without "_a" like "gm"?

**Review by hx89 — APPROVED** [review]

File: torch/quantization/ns/graph_matcher.py (hunk 2 above)
Comment (vkuzo) [reply torch/quantization/ns/graph_matcher.py:290]: sure, let me do that in a future PR (soon)
